tla/consensus/ccfraft.tla

```diff
@@ -941,6 +941,9 @@ HandleAppendEntriesResponse(i, j, m) ==
           \* UNCHANGED matchIndex is implied by the following statement in figure 2, page 4 in the raft paper:
            \* "If AppendEntries fails because of log inconsistency: decrement nextIndex (aka sentIndex +1) and retry"
           /\ UNCHANGED matchIndex
+       \/ /\ m.term < currentTerm[i]
+          /\ m.success
+          /\ UNCHANGED <<matchIndex, sentIndex>>
```

Comment:
@lemmy @heidihoward When we're matching a `recv_append_entries_response` line, we have to match the case where it's an ACK (success == true), but from an old term, so ignored/discarded.

I've stuffed it in here, to benefit from the surrounding `UNCHANGED`, rather than adding a `\/ Discard(m) /\ UNCHANGED << >>` to `Traceccfraft`'s `IsRcvAppendEntriesResponse`. I'm sure this isn't the right place, but it seems to work.

How does `IsRcvRequestVoteResponse`, or indeed any other recv match, handle this?

(NB: This is almost exactly what we're discussing in #5927 - AERs get special-case handling of their terms)

Reply:
I have tried to make a similar change in #5941, but only for the spec. I don't think I understand how this works, because it seem to me that the level above (RcvAppendEntriesResponse) still allows non-deterministic drops, and so will allow traces that drop stale NACKs.

Reply:
(I think, with caveats*) Non-deterministic drops are still allowed - they produce a state equivalent to if the message had been dropped in the network, albeit partway through executing an action. A change in `RcvAppendEntriesResponse` doesn't help us with trace validation, because the failure here is via `IsRcvAppendEntriesResponse`, which calls `HandleAppendEntriesResponse` directly.

So either we need to catch it here which is considered by both paths, or maybe the change you've made in `RcvAppendEntriesResponse` could be in both `RcvAppendEntriesResponse` and (the similarly structured) `IsRcvAppendEntriesResponse`? I can't parse what the semantic difference is, or which is preferable. The trace spec also has `RcvAppendEntriesResponse` in its `ComposedNext`, but I don't understand when that is called.

Reply:
> Non-deterministic drops are still allowed - they produce a state equivalent to if the message had been dropped in the network, albeit partway through executing an action

I do agree, but I think it's clearer for the network-induced dropping to be done in a single explicit location (DropMessagesTo). Here we are effectively allowing one special case of dropping and bundling it with a real implementation behaviour (dropping stale ACKs) under "DropStaleResponses", which don't accurately describe the reason we're dropping the message in the NACK cases.

Yes I also think doing it one level above means duplicating it, because the trace has somewhat different logic for handling append entries unfortunately.

Anyway, I think I'm happy with either, given a comment.